Add db::create_copy and factor algorithm code

nilsdeppe opened the pull request:
## Proposed changes

- Want to be able to get a non-aliasing copy of a DataBox in the tests sometimes when testing Actions on multiple parallel components
- Factor some of the Algorithm helper functions so they can be used by the mocking framework.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add type trait for can_be_copy_constructed

The std::is_copy_constructible does not work as expected with std::unordered_map

nilsdeppe committed:
Add Deferred deep_copy() and evaluated()

- deep_copy() creates a non-aliasing copy of a Deferred that is not a lazy
  function. Cannot do this for lazy functions since the arguments may need to be
  deep copied too if they're Deferreds.
- evaluated() returns true if the lazy function is evaluated, false if
  not. Returns true for object Deferreds, ie not a lazy function.

nilsdeppe committed:
Add db::create_copy that creates a non-aliasing copy

create_from creates an aliasing DataBox where if the original DataBox is mutated
so is the new copy.

create_copy creates a non-aliasing DataBox where if the original DataBox is
mutated the new won't be.

nilsdeppe committed:
Factor out metafunctions from Algorithm.hpp

- We want to be able to reuse the metafunctions in the mocking framework without
  having to include all of Algorithm.hpp

nilsdeppe committed:
Remove duplicates in BoostHelpers make_boost_variant_over

nilsdeppe committed:
Factor out simple action visitation from Algorithm.hpp

We want to be able to reuse as much code from the Algorithm implementation in
the mocking framework as possible. Visitation is one of the things we want to
use.

wthrowe commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -373,6 +373,29 @@ using remove_cvref_t = typename remove_cvref<T>::type;
+ * \brief Check if `T` is copy constructible
+ *
+ * The STL `std::is_copy_constructible` does not work correctly with some types,
+ * such as `std::unordered_map` (likely a problem with `std::unordered_map`).
```
`std::is_copy_constructible` doesn't do what you want (which would generally be the more useful behavior), but it does what the standard says it should do (as do the containers).  Please name this something else to avoid confusion and document the difference from `std::is_copy_constructible`.

nilsdeppe replied:
Do you have a suggestion for an alternative name? I've be trying to think of something that doesn't suck and it's been a struggle. Maybe `can_be_copy_constructed`?

wthrowe replied:
I can't think of anything better.

wthrowe commented on tests/Unit/DataStructures/DataBox/Test_DataBox.cpp:
```diff
@@ -1635,6 +1635,185 @@ SPECTRE_TEST_CASE("Unit.DataStructures.DataBox.Subitems",
+}
+
+// [[OutputRegex, Cannot make a deep copy of a Deferred that holds a
+// non-copyable data member.]]
```
This is a user-facing error, so can it say what is wrong with the operation in terms of the DataBox instead of in terms of Deferred, which most users won't have a clue about?  I wouldn't understand this error message if I hadn't just reviewed the code causing it, and I'd still have to read the code again to figure out which of the two items in the box it was complaining about.

nilsdeppe replied:
100000% agreed :)

wthrowe commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1305,6 +1343,10 @@ SPECTRE_ALWAYS_INLINE constexpr auto create_from(Box&& box,
  *
+ * Note that in the rvalue and const lvalue case the output DataBox shares all
+ * items that were not removed with the input DataBox. This means if an item is
+ * mutated in the input DataBox it is also mutated in the output DataBox.
```
Isn't this true in every case, not just rvalue and const lvalue?  I guess those are the only "real" cases now (the others are implicitly converted to const lvalue), but this reads as if it is contrasting with other types, particularly when read immediately after the previous paragraph.  Also, this isn't a terribly meaningful statement for the rvalue case, since the input DataBox has been moved out of so mutating items in it is not a meaningful operation.

nilsdeppe replied:
Sorry, I think what I was trying to convey is that in the `const lvalue` case there's aliasing, but this was written before deciding on a name for `create_copy`. I'll clarify :)

nilsdeppe commented:
I've:
- Rebased on develop after latest merges
- Addressed the comments using `can_be_copy_constructed` for now, but I'm happy to change that if a different name is desired :)

kidder reviewed (commented):
I am happy with the fixups

nilsdeppe commented:
- Sorted includes in `Algorithm.hpp`

wthrowe reviewed (dismissed):
I'm happy.  Squash.

nilsdeppe commented:
Alright, squashed :)

nilsdeppe commented:
Travis failures are timeouts, all tests passed.
